### PR TITLE
server: bind port lazily when it's configured

### LINKED ISSUE #2999: Bind port only right before server start
Currently the listen port is bind long before start, so that checking TiKV running according to port is not reliable. We need to bind the port right before the server starts.

### LINKED ISSUE #4224: raftstore panic when staring TiKV
## Bug Report

**What did you do?**

Just start TiKV, panic. but after panic several times, the TiKV can start successfully. 

**What did you expect to see?**

No panic

**What did you see instead?**

```
[2019/02/16 20:46:32.466 +08:00] [ERROR] [mod.rs:475] ["thread 'raftstore-6-0' panicked 'called `Option::unwrap()` on a `None` value' at \"src/libcore/option.rs:355\"\nstack backtrace:\n   0:     0x55b0eda0898d - backtrace::backtrace::libunwind::trace::h42bd35238981daad\n                        at /home/qupeng/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/libunwind.rs:54\n                         - backtrace::backtrace::trace::h1f00fe0b81b1621d\n                        at /home/qupeng/.cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.2.3/src/backtrace/mod.rs:70\n   1:     0x55b0ed38c75b - tikv::util::set_panic_hook::{{closure}}::h9c9bb177cf30c18e\n                        at /home/qupeng/github.com/pingcap/tikv/target/release/build/backtrace-ce050514236e0111/out/capture.rs:79\n   2:     0x55b0edaed9fd - std::panicking::rust_panic_with_hook::he9362bdf637fc5d8\n                        at src/libstd/panicking.rs:495\n   3:     0x55b0edaed7be - std::panicking::continue_panic_fmt::h8c8ecd2074b327b3\n                        at src/libstd/panicking.rs:398\n   4:     0x55b0edb01ef5 - rust_begin_unwind\n                        at src/libstd/panicking.rs:325\n   5:     0x55b0edb0685b - core::panicking::panic_fmt::hc84bf3901c1095d4\n                        at src/libcore/panicking.rs:95\n   6:     0x55b0edb0831a - core::panicking::panic::h2cd043862cc6dc4a\n                        at src/libcore/panicking.rs:59\n   7:     0x55b0ed26bd46 - <core::option::Option<T>>::unwrap::h6f447e9a70a6d3cf\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libcore/macros.rs:20\n                         - <tikv::raftstore::store::fsm::store::StoreFsmDelegate<'a, T, C>>::store_heartbeat_pd::h14cd37177d1aa0fd\n                        at /home/qupeng/github.com/pingcap/tikv/src/raftstore/store/fsm/store.rs:1534\n   8:     0x55b0ed2631ef - <tikv::raftstore::store::fsm::store::StoreFsmDelegate<'a, T, C>>::handle_msgs::h1cb2dce6bf1dd9be\n                        at /home/qupeng/github.com/pingcap/tikv/src/raftstore/store/fsm/store.rs:452\n   9:     0x55b0ed1e944a - <tikv::raftstore::store::fsm::store::RaftPoller<T, C> as tikv::raftstore::store::fsm::batch::PollHandler<tikv::raftstore::store::fsm::peer::PeerFsm, tikv::raftstore::store::fsm::store::StoreFsm>>::handle_control::h953defd5ec158028\n                        at /home/qupeng/github.com/pingcap/tikv/src/raftstore/store/fsm/store.rs:615\n                         - <tikv::raftstore::store::fsm::batch::Poller<N, C, Handler>>::poll::hfc9fcd194235867e\n                        at /home/qupeng/github.com/pingcap/tikv/src/raftstore/store/fsm/batch.rs:334\n  10:     0x55b0ed007153 - <tikv::raftstore::store::fsm::batch::BatchSystem<N, C>>::spawn::{{closure}}::h4216ac29e8293fd7\n                        at /home/qupeng/github.com/pingcap/tikv/src/raftstore/store/fsm/batch.rs:412\n                         - std::sys_common::backtrace::__rust_begin_short_backtrace::h88070bb57e05ea4a\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/sys_common/backtrace.rs:136\n  11:     0x55b0ed14ff20 - std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}::hfd8153758c5de686\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/thread/mod.rs:477\n                         - <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once::hf66778517c33afb3\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/panic.rs:319\n                         - std::panicking::try::do_call::h9fc881c74d130201\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/panicking.rs:310\n                         - std::panicking::try::h32753ac12e5d4ef1\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56//src/libpanic_unwind/lib.rs:102\n                         - std::panic::catch_unwind::h39e4a909b32d19ce\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/panic.rs:398\n                         - std::thread::Builder::spawn_unchecked::{{closure}}::h7c9247b83564e7d6\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/libstd/thread/mod.rs:476\n                         - <F as alloc::boxed::FnBox<A>>::call_box::h74fc0ea5bd0b31f6\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/liballoc/boxed.rs:673\n  12:     0x55b0edb03a9d - <alloc::boxed::Box<(dyn alloc::boxed::FnBox<A, Output=R> + 'a)> as core::ops::function::FnOnce<A>>::call_once::hb8f4ee9bdf46e5d9\n                        at /rustc/14997d56a550f4aa99fe737593cd2758227afc56/src/liballoc/boxed.rs:683\n                         - std::sys_common::thread::start_thread::h938451c06a28c9dd\n                        at src/libstd/sys_common/thread.rs:24\n                         - std::sys::unix::thread::Thread::new::thread_start::h809dab03f76374f4\n                        at src/libstd/sys/unix/thread.rs:90\n  13:     0x7f202c7b6dd4 - start_thread\n  14:     0x7f202bbb6eac - __clone\n  15:                0x0 - <unknown>"]
[2019/02/16 20:46:32.466 +08:00] [INFO] [unknown:920] ["[region 3201491] 3201493 [term: 8] received a MsgHeartbeat message with higher term from 3201492 [term: 9]"]
```

### PR DESCRIPTION
**Pull request by BusyJay**
## What have you changed? (mandatory)

This PR make TiKV only listens to given port when it's started. Note that listening to port before everything is prepared is a mistake and can lead to unexpected bugs like #4224.

Binding port before server is started is required by our tests to avoid port conflict. So this PR delays binding port only when it's configured.

## What are the type of the changes? (mandatory)

- Bug fix (change which fixes an issue)

## How has this PR been tested? (mandatory)

manual tests

## Does this PR affect documentation (docs) update? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

Close #2999
Close #4224

### CODE REVIEW

**Comment by zhangjinpeng1987** [comment]
Is there any integration test to cover https://github.com/tikv/tikv/issues/4224?

**Comment by BusyJay** [comment]
#4224 is expected because messages should not be sent before raftstore is started.

File: src/server/server.rs [hunk 1]
```diff
@@ -129,12 +132,14 @@ impl<T: RaftStoreRouter, S: StoreAddrResolver + 'static> Server<T, S> {
             if let Some(service) = import_service {
                 sb = sb.register_service(create_import_sst(service));
             }
-            sb.build()?
-        };
-
-        let addr = {
-            let (ref host, port) = grpc_server.bind_addrs()[0];
-            SocketAddr::new(IpAddr::from_str(host)?, port as u16)
+            if addr.port() == 0 {
```
Comment (siddontang) [inline src/server/server.rs]: please add comments for port 0 here.

**Pull request merged**